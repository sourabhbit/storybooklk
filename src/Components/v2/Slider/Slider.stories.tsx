import { ComponentMeta, ComponentStory } from '@storybook/react'
import { generateArgTypes } from 'Helpers/Storybook'
import Slider from './index'

export default {
	title: 'Components/v2/Slider',
	component: Slider,
	// decorators: [withKnobs],
	argTypes: {
		id: generateArgTypes({
			type: 'string',
			required: true,
			description: 'String used as unique identifier',
			category: 'General',
		}),
		startValue: {
			...generateArgTypes({
				type: 'string',
				required: false,
				description: 'starting value of price range',
				category: 'Component',
			}),
		},
		endValue: generateArgTypes({
			type: 'string',
			required: false,
			description: 'end value of price range',
			category: 'Component',
		}),
		minDiff: generateArgTypes({
			type: 'number',
			required: false,
			description: 'minimum difference between starting and end value',
			category: 'Component',
		}),
		maxRangeValue: generateArgTypes({
			type: 'number',
			required: false,
			description: 'Maximum value of a product',
			category: 'Component',
		}),
		step: generateArgTypes({
			type: 'number',
			required: false,
			description: 'step of increment/decrement',
			category: 'Component',
		}),
		onInput: {
			...generateArgTypes({
				type: 'function',
				required: false,
				description: 'onInput handler for slider',
				category: 'Events',
			}),
		},
	},
} as ComponentMeta<typeof Slider>

const Template: ComponentStory<typeof Slider> = (args) => <Slider {...args} />

export const Overview = Template.bind({})
Overview.args = {
	id: 'slider',
	startValue: '500',
	endValue: '3000',
	minDiff: 1000,
	maxRangeValue: 10000,
	step: 100,
}
