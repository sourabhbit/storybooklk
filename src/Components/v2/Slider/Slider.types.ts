export interface SliderTypes {
	id: string
	startValue: string
	endValue: string
	minDiff?: number
	maxRangeValue: number
	step?: number
	onInput?: () => void
}
