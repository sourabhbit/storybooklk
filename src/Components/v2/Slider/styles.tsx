import styled from 'styled-components'

export const Container = styled.div`
	position: relative;
	height: 2px;
	outline: none;
`
export const SliderTrack = styled.div<{
	percentOne: number
	percentTwo: number
}>`
	height: 100%;
	margin: auto;
	top: 0;
	bottom: 0;
	background: ${(props) =>
		`linear-gradient(to right, var(--warm-grey-100) ${props.percentOne}% , var(--dark-blue-100) ${props.percentOne}% , var(--dark-blue-100) ${props.percentTwo}%, var(--warm-grey-100) ${props.percentTwo}%)`};
`

export const SliderInput = styled.input.attrs({
	type: 'range',
})`
	position: absolute;
	top: 0;
	bottom: 0;
	width: 100%;
	-webkit-appearance: none;
	pointer-events: none;
	background: none;
	outline: none;
	::-webkit-slider-thumb {
		pointer-events: auto;
		-webkit-appearance: none;
		width: 57px;
		height: 34px;
		cursor: pointer;
	}
`
export const PriceWrapper = styled.div<{
	left: number
	selected: boolean
}>`
	position: absolute;
	z-index: ${(props) => (props.selected ? 2 : 1)};
	display: flex;
	justify-content: center;
	align-items: center;
	top: -16px;
	bottom: 0;
	width: 57px;
	height: 34px;
	border: 1px solid var(--dark-blue-100);
	border-radius: var(--border-radius-xs);
	background: var(--white);
	pointer-events: none;
	left: ${(props) => `${props.left}%`};
	transform: ${(props) => `translate(-${props.left}%, 0)`};
	::before {
		transform: ${(props) => `translate(-${props.left}%, 0)`};
	}
`

export const InputWrapper = styled.div``
