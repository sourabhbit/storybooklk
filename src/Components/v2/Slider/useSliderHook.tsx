import { action } from '@storybook/addon-actions'
import { useCallback, useState } from 'react'
// import { SliderTypes } from './Slider.types'

interface sliderTypes {
	startValue: string
	endValue: string
	maxRangeValue: number
	minDiff: number
}
export const useSliderHook = ({
	startValue,
	endValue,
	maxRangeValue,
	minDiff,
}: sliderTypes) => {
	const [minValue, setMinValue] = useState<string>(startValue)
	const [maxValue, setMaxValue] = useState<string>(endValue)
	const [percentOne, setPercentOne] = useState<number>(
		(parseInt(minValue) / maxRangeValue) * 100
	)
	const [percentTwo, setPercentTwo] = useState<number>(
		(parseInt(maxValue) / maxRangeValue) * 100
	)
	const fillColor = useCallback(() => {
		setPercentOne((parseInt(minValue) / maxRangeValue) * 100)
		setPercentTwo((parseInt(maxValue) / maxRangeValue) * 100)
	}, [minValue, maxValue, maxRangeValue])
	const slideOne = (e: React.ChangeEvent<HTMLInputElement>) => {
		const value = e.target.value
		if (parseInt(maxValue) - parseInt(value) <= minDiff) {
			setMinValue((parseInt(maxValue) - minDiff).toString())
		} else {
			setMinValue(value)
			action('start value changed')(e, value)
		}
	}
	const slideTwo = (e: React.ChangeEvent<HTMLInputElement>) => {
		const value = e.target.value
		if (parseInt(value) - parseInt(minValue) <= minDiff) {
			setMaxValue((parseInt(minValue) + minDiff).toString())
		} else {
			setMaxValue(value)
			action('end value changed')(e, value)
		}
	}

	return {
		minValue,
		maxValue,
		percentOne,
		percentTwo,
		fillColor,
		slideOne,
		slideTwo,
	}
}
