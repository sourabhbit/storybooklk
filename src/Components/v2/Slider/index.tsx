import { SimplePrice } from 'Components/v2/Price/Price.stories'
import React, { useEffect, useState } from 'react'
import { ComponentSizeENUM, TypographyENUM } from 'Types/general'
import { SliderTypes } from './Slider.types'
import { Container, SliderTrack, SliderInput, PriceWrapper } from './styles'
import { useSliderHook } from './useSliderHook'

const Slider = ({
	id,
	startValue = '500',
	endValue = '3000',
	minDiff = 1000,
	maxRangeValue = 10000,
	step = 500,
}: SliderTypes) => {
	const [selectedOne, setSelectedOne] = useState(true)
	const [selectedTwo, setSelectedTwo] = useState(false)
	const {
		minValue,
		maxValue,
		percentOne,
		percentTwo,
		fillColor,
		slideOne,
		slideTwo,
	} = useSliderHook({ startValue, endValue, maxRangeValue, minDiff })
	useEffect(() => {
		fillColor()
	}, [fillColor, startValue, endValue])

	return (
		<Container id={id} data-testid="slider-test">
			<SliderTrack
				percentOne={percentOne}
				percentTwo={percentTwo}
			></SliderTrack>
			<label htmlFor="slider-input">
				<SliderInput
					min="0"
					max={maxRangeValue}
					value={minValue}
					step={step}
					onChange={(e) => slideOne(e)}
					onMouseDown={() => {
						setSelectedOne(true)
						setSelectedTwo(false)
					}}
				/>
			</label>
			<PriceWrapper left={percentOne} selected={selectedOne}>
				<SimplePrice
					id={id}
					amount={minValue.toString()}
					special={false}
					font={TypographyENUM.serif}
					componentSize={ComponentSizeENUM.small}
				/>
			</PriceWrapper>
			<label htmlFor="slider-input">
				<SliderInput
					min="0"
					max={maxRangeValue}
					value={maxValue}
					step={step}
					onChange={(e) => slideTwo(e)}
					onMouseDown={() => {
						setSelectedOne(false)
						setSelectedTwo(true)
					}}
				/>
			</label>
			<PriceWrapper left={percentTwo} selected={selectedTwo}>
				<SimplePrice
					id={id}
					amount={maxValue.toString()}
					special={false}
					font={TypographyENUM.serif}
					componentSize={ComponentSizeENUM.small}
				/>
			</PriceWrapper>
		</Container>
	)
}

export default Slider
