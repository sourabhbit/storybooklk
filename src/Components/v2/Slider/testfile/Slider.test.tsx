import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import Slider from '../index'

const props = {
	id: 'Slider',
	startValue: '200',
	endValue: '700',
	maxRangeValue: 3000,
}

describe('<Slider />', () => {
	let _asFragment: any = null

	beforeEach(() => {
		const { asFragment } = render(<Slider {...props} />)
		_asFragment = asFragment
	})

	afterAll(() => {
		_asFragment = null
	})

	it('Should render component', () => {
		const container = screen.getByTestId('slider-test')
		expect(container).toBeInTheDocument()
	})

	it('Should take a snapshot', () => {
		expect(_asFragment()).toMatchSnapshot()
	})
})
