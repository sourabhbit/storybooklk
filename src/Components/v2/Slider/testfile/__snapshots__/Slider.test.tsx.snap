// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Slider /> Should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-dkPtRN brQDgy"
    data-testid="slider-test"
    id="Slider"
  >
    <div
      class="sc-hKwDye iivCqM"
    />
    <label
      for="slider-input"
    >
      <input
        class="sc-eCImPb qzrAe"
        max="3000"
        min="0"
        step="500"
        type="range"
        value="200"
      />
    </label>
    <div
      class="sc-jRQBWg fBseAg"
    >
      <span
        class="sc-bdvvtL fOQrUR"
        data-testid="pricing-test"
      >
        200
        <span
          class="sc-gsDKAQ logYvF"
          color="var(--warm-grey-50)"
        />
      </span>
    </div>
    <label
      for="slider-input"
    >
      <input
        class="sc-eCImPb qzrAe"
        max="3000"
        min="0"
        step="500"
        type="range"
        value="700"
      />
    </label>
    <div
      class="sc-jRQBWg gBhjUx"
    >
      <span
        class="sc-bdvvtL fOQrUR"
        data-testid="pricing-test"
      >
        700
        <span
          class="sc-gsDKAQ logYvF"
          color="var(--warm-grey-50)"
        />
      </span>
    </div>
  </div>
</DocumentFragment>
`;
